log/src/io/pedestal/log.clj

```diff
@@ -374,14 +374,14 @@
   Note:
   If you mix `with-context` with the more basic `with-context-kv`, you may see undesired keys/values in the log"
   [ctx-map & body]
-  (let [formatter (::formatter ctx-map pr-str)]
-    (if (empty? ctx-map) ;; Optimize for the code-gen/dynamic case where the map may be empty
-      `(do
-         ~@body)
-      `(let [old-ctx# *mdc-context*
-             mdc# (or ~(::mdc ctx-map) (MDC/getMDCAdapter))]
-         (binding [*mdc-context* (merge *mdc-context* ~ctx-map)]
-           (-put-mdc mdc# mdc-context-key (~formatter (dissoc *mdc-context*
+  `(let [ctx-map-val# ~ctx-map]
+     (if (empty? ctx-map-val#) ;; Optimize for the code-gen/dynamic case where the map may be empty
```

Comment:
I wonder how often this `(empty? ctx-map)` predicate is satisfied, especially in the previous impl when you had to explicitly pass `{}` for it to trigger. We could preserve this optimization at expansion time but it would complicate the implementation a bit with conditionals

Reply:
`(empty? nil)` returns true as well.  This portion of the logging library expects that most users will build a higher-level context construct on top of these calls (their own macro for sourcing and setting the MDC).
I think checking for `map?` is probably the best approach here, reducing the branching within the macro's generated code.

As a general rule of thumb, the logging utilities attempt to remove as much overhead from the logging interactions as possible.  It's common in high-throughput scenarios to see GC throughput limit application goodput -- one source of this can be excessive logging, forcing developers to decide between diagnostics and operational requirements.

Pedestal2 also introduces `with-mdc-copy` which avoids the `formatter` pieces and works against the MDC directly, correctly copying the MDC across thread boundaries.  In the systems I've been working with, this approach is more common because it's friendlier for structured log processors (like DataDog, etc.)

Reply:
Cool, thanks for the context, it really helps me wrap my head around whats going/the constraints in flight.

> I think checking for map? is probably the best approach here, reducing the branching within the macro's generated code.

so essentially just go with https://github.com/pedestal/pedestal/pull/713/commits/806bca8f4f17bcd16cf236c5fb310487167301cb for a fix? And you think it is fine to then fail and fall back with the `::formatter` and `::mdc` lookups?

Reply:
Sorry for the confusion -- you want `nil` to short-circuit (as it already does):
If a developer's higher-level macro produced a `nil` context, you want this to fall through the correct code-gen path.

Reply:
revisiting this now. How does https://github.com/pedestal/pedestal/pull/713/commits/6a028e78d5cc4bd36fb6e5eeb31e3aae3f6f3ce2 look?